lib/spack/spack/test/stage.py

```diff
@@ -411,23 +412,35 @@ def check_stage_dir_perms(prefix, path):
     # Ensure the path's subdirectories -- to `$user` -- have their parent's
     # perms while those from `$user` on are owned and restricted to the
     # user.
-    status = os.stat(prefix)
-    gid = status.st_gid
-    uid = os.getuid()
+    assert path.startswith(prefix)
+
     user = getpass.getuser()
-    parts = path[len(prefix) + 1:].split(os.path.sep)
-    have_user = False
-    for part in parts:
-        if part == user:
-            have_user = True
-        prefix = os.path.join(prefix, part)
-        prefix_status = os.stat(prefix)
-        if not have_user:
-            assert gid == prefix_status.st_gid
-            assert status.st_mode == prefix_status.st_mode
-        else:
-            assert uid == status.st_uid
-            assert status.st_mode & stat.S_IRWXU == stat.S_IRWXU
+    prefix_status = os.stat(prefix)
+    uid = os.getuid()
+
+    # Obtain lists of ancestor and descendant paths of the $user node, if any.
+    #
+    # Skip processing prefix ancestors since no guarantee they will be in the
+    # required group (e.g. $TEMPDIR on HPC machines).
+    skip = prefix if prefix.endswith(os.sep) else os.path.join(prefix, os.sep)
+    group_paths, user_node, user_paths = partition_path(path.replace(skip, ""),
+                                                        user)
+
+    for p in group_paths:
+        parent = os.path.dirname(p)
```

Comment:
Invoking `partition_path` on a path segments it into a series of directories that leads up to a file. Therefore, I think it is confusing to invoke `dirname` on any entry in `group_paths`: for example from `test_partition_path` there is the case of `partition_path('relative/path', 'path') == (['relative'], 'relative/path', [])` in which case `group_paths == ['relative']`; `relative` is a directory in this case but invoking `dirname` on it yields `''`.

Reply:
Great catch.  Thanks.